lldb/include/lldb/Core/DILAST.h

```diff
@@ -0,0 +1,459 @@
+
+  virtual bool is_rvalue() const = 0;
+  virtual bool is_bitfield() const { return false; };
+  virtual bool is_context_var() const { return false; };
+  virtual bool is_literal_zero() const { return false; }
+  virtual uint32_t bitfield_size() const { return 0; }
+  virtual CompilerType result_type() const = 0;
+
+  clang::SourceLocation GetLocation() const { return m_location; }
+  NodeKind getKind() const { return m_kind; }
+
+  // The expression result type, but dereferenced in case it's a reference. This
+  // is for convenience, since for the purposes of the semantic analysis only
+  // the dereferenced type matters.
+  CompilerType GetDereferencedResultType() const;
+
+private:
+  clang::SourceLocation m_location;
+  const NodeKind m_kind;
+};
+
+using DILASTNodeUP = std::unique_ptr<DILASTNode>;
+
+class ErrorNode : public DILASTNode {
+public:
+  ErrorNode(CompilerType empty_type)
+      : DILASTNode(clang::SourceLocation(), NodeKind::eErrorNode),
+        m_empty_type(empty_type) {}
+  void Accept(Visitor *v) const override;
+  bool is_rvalue() const override { return false; }
+  CompilerType result_type() const override { return m_empty_type; }
+  CompilerType result_type_real() const { return m_empty_type; }
+
+  static bool classof(const DILASTNode *node) {
+    return node->getKind() == NodeKind::eErrorNode;
+  }
+
+private:
+  CompilerType m_empty_type;
+};
+
+class ScalarLiteralNode : public DILASTNode {
+public:
+  ScalarLiteralNode(clang::SourceLocation location, CompilerType type,
+                    Scalar value, bool is_literal_zero)
+      : DILASTNode(location, NodeKind::eScalarLiteralNode), m_type(type),
+        m_value(value), m_is_literal_zero(is_literal_zero) {}
+
+  void Accept(Visitor *v) const override;
+  bool is_rvalue() const override { return true; }
+  bool is_literal_zero() const override { return m_is_literal_zero; }
+  CompilerType result_type() const override { return m_type; }
+
+  auto value() const { return m_value; }
+
+  static bool classof(const DILASTNode *node) {
+    return node->getKind() == NodeKind::eScalarLiteralNode;
+  }
+
+private:
+  CompilerType m_type;
+  Scalar m_value;
+  bool m_is_literal_zero;
+};
+
+class StringLiteralNode : public DILASTNode {
+public:
+  StringLiteralNode(clang::SourceLocation location, CompilerType type,
+                    std::vector<char> value, bool is_literal_zero)
+      : DILASTNode(location, NodeKind::eStringLiteralNode), m_type(type),
+        m_value(value), m_is_literal_zero(is_literal_zero) {}
+
+  void Accept(Visitor *v) const override;
+  bool is_rvalue() const override { return true; }
+  bool is_literal_zero() const override { return m_is_literal_zero; }
+  CompilerType result_type() const override { return m_type; }
+
+  auto value() const { return m_value; }
+
+  static bool classof(const DILASTNode *node) {
+    return node->getKind() == NodeKind::eStringLiteralNode;
+  }
+
+private:
+  CompilerType m_type;
+  std::vector<char> m_value;
```

Comment:
It sort of makes more sense for a string literal to be an array (or vector) of characters.  But I could make this a std::string if you really want me to?

Reply:
A string *is* a vector of characters, but with extra operations that make sense for operating on strings. If I see vector<char>, I start to think whether there's something subtle going on, like maybe this representing a raw memory dump of something. While it's not possible to make a direct comparison to clang's StringLiteral class (because it uses tail allocations to store the string), its interface uses `StringRef` objects (and not `ArrayRef<char>`), so yes, I think this should be a string.